From 731bffe936c3dcf0712b67c28b1724e3057a2b77 Mon Sep 17 00:00:00 2001
From: Evan Wang <xswang@marvell.com>
Date: Tue, 22 Nov 2016 17:15:49 +0800
Subject: [PATCH 15/45] u3d: add valid check before schedule workqueue

- Schedule gadget glue workqueue in U3D driver without
  validation might cause kernel crash, since it is connected
  to U3D driver in routine udc_detect, and it is not safe/correct
  to assume it has always been invoked already.
- There was crash while gadget glue was not involved in the configfs.
  Back then, if U3D driver is bind to configfs, while a USB2 Host is
  connected to Armada3700 DB, then U3D driver gets a interrupt about
  "USB2 Connection Status", and try to schedule the workqueue which
  has not been initialized, and got crash.
- But since now gadget glue has already been involved to configfs with
  patch "usb: gadget: automatically select correct/working udc driver
  for configfs", this crash would not occur.
- The patch adds valid check of glue layer workqueue before
  U3D driver schedule it.

Change-Id: I70a15db219bd5ac0d6572a354cd0a7e0d7be3d90
Signed-off-by: Evan Wang <xswang@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/33946
Reviewed-by: Omri Itach <omrii@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Tested-by: Nadav Haklai <nadavh@marvell.com>
---
 drivers/usb/gadget/udc/mvebu_u3d.c | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/drivers/usb/gadget/udc/mvebu_u3d.c b/drivers/usb/gadget/udc/mvebu_u3d.c
index 263563873776..9a64928314e4 100644
--- a/drivers/usb/gadget/udc/mvebu_u3d.c
+++ b/drivers/usb/gadget/udc/mvebu_u3d.c
@@ -1960,7 +1960,8 @@ static irqreturn_t mvc2_irq(int irq, void *devid)
 
 				cp->status &= ~MVCP_STATUS_USB2;
 				glue.status = cp->status;
-				schedule_work(cp->work);
+				if (cp->work)
+					schedule_work(cp->work);
 			}
 		}
 
@@ -1983,7 +1984,8 @@ static irqreturn_t mvc2_irq(int irq, void *devid)
 
 			cp->status |= MVCP_STATUS_USB2;
 			glue.status = cp->status;
-			schedule_work(cp->work);
+			if (cp->work)
+				schedule_work(cp->work);
 		}
 
 		if ((refint & MVCP_REF_INTEN_USB2_DISCNT) &&
@@ -1993,7 +1995,8 @@ static irqreturn_t mvc2_irq(int irq, void *devid)
 			if (mvc2_checkvbus(cp)) {
 				cp->status &= ~MVCP_STATUS_USB2;
 				glue.status = cp->status;
-				schedule_work(cp->work);
+				if (cp->work)
+					schedule_work(cp->work);
 			}
 		}
 
-- 
2.35.3

